src/attestation_api.nim

```diff
@@ -0,0 +1,411 @@
+      warn("Code signing not initialized. Run `chalk setup` to fix.")
+      # there is no key in self chalkmark
+      return
+
+    let provider = getProvider()
+    try:
+      provider.init(provider)
+    except:
+      error("Attestation key provider is misconfigured: " & getCurrentExceptionMsg())
+      return
+
+    if withPrivateKey:
+      try:
+        cosignKey.password = provider.retrievePassword(provider, cosignKey)
+      except:
+        error("Could not retrieve cosign private key password: " & getCurrentExceptionMsg())
+        return
+
+      try:
+        if not cosignKey.isValid():
+          cosignKey = AttestationKey(nil)
+          return
+      except:
+        cosignKey = AttestationKey(nil)
+        return
+
+proc saveKeyToSelf(key: AttestationKey): bool =
+  let selfChalk = getSelfExtraction().get()
+  selfChalk.extract["$CHALK_ENCRYPTED_PRIVATE_KEY"] = pack(key.privateKey)
+  selfChalk.extract["$CHALK_PUBLIC_KEY"]            = pack(key.publicKey)
+
+  let savedCommandName = getCommandName()
+  setCommandName("setup")
+  try:
+    result = selfChalk.writeSelfConfig()
+  finally:
+    setCommandName(savedCommandName)
+
+proc setupAttestation*() =
+  info("Ensuring cosign is present to setup attestation.")
+  try:
+    if getCosignLocation(downloadCosign = true) == "":
+      raise newException(ValueError, "Failed to get cosign binary")
+  except:
+    raise newException(ValueError, "Failed to get cosign binary: " & getCurrentExceptionMsg())
+
+  let provider = getProvider()
+  try:
+    provider.init(provider)
+  except:
+    raise newException(ValueError, "Attestation key provider is misconfigured: " & getCurrentExceptionMsg())
+
+  # a bit of nesting of exceptions to propagate appropriate error
+  # to the user as key providers can optionally implmenet
+  # retrieval/generation
+  try:
+    if provider.retrieveKey != nil:
+      cosignKey = provider.retrieveKey(provider)
+      if cosignKey == nil:
+        raise newException(ValueError, "Retrieved invalid key")
+    else:
+        raise newException(ValueError, "Existing key retrieval not supported")
+  except:
+    if provider.generateKey == nil:
+      raise newException(ValueError, "Provider '" & provider.name & "': " & getCurrentExceptionMsg())
+    try:
+      cosignKey = provider.generateKey(provider)
+      if cosignKey == nil:
+        raise newException(ValueError, "Generated invalid key")
+    except:
+      raise newException(ValueError, "Provider '" & provider.name & "': " & getCurrentExceptionMsg())
+
+  try:
+    if not cosignKey.saveKeyToSelf():
+      raise newException(ValueError, "Failed to store generated attestation key to chalk")
+  except:
+    raise newException(ValueError, "Failed to store generated attestation key to chalk: " & getCurrentExceptionMsg())
+
+# ----------------------------------------------------------------------------
+# Everything below is related to actually signing/validating an artifact
```

Comment:
everything below hasnt changes except some minor refactor with function signature changes/etc. the key provider changes are all above in this file